### PR TITLE
Factorization machine example & sparse example folder re-org

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##

Added an example for factorization machine, and created separate folders for the examples

@ZiyueHuang @anirudh2290  pls help review

### Essentials ###
- [ ] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] For user-facing API changes, API doc string has been updated. For new C++ functions in header files, their functionalities and arguments are well-documented. 
- [ ] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: example/sparse/factorization_machine/metric.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import mxnet as mx
+import numpy as np
+
+@mx.metric.register
+@mx.metric.alias('log_loss')
+class LogLossMetric(mx.metric.EvalMetric):
```
Comment (eric-haibin-lin) [inline example/sparse/factorization_machine/metric.py]: The existing `nll_loss` metric was not used because it was expecting K-D output. For FM the output is 1-D for binary classification

File: example/sparse/factorization_machine/data.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    def next(self):
+        return self.the_batch
+
+
+def get_criteo_data(data_dir):
+    if not os.path.isdir(data_dir):
+        os.mkdir(data_dir)
+    try:
+        logging.info("Downloading dataset criteo to " + data_dir + " now ...")
+        os.system("aws s3 cp --recursive --no-sign-request s3://sparse-dataset/criteo " + data_dir)
```
Comment (anirudh2290) [inline example/sparse/factorization_machine/data.py]: How big is this directory. Maybe add a warning about the size here?

File: example/sparse/factorization_machine/train.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    train_data = mx.io.LibSVMIter(data_libsvm=args.data, data_shape=(num_features,),
+                                  batch_size=batch_size)
+    # model
+    model = factorization_machine_model(factor_size, num_features)
+
+    # module
+    mod = mx.mod.Module(symbol=model)
+    mod.bind(data_shapes=train_data.provide_data, label_shapes=train_data.provide_label)
+    mod.init_params()
+    optimizer_params=(('learning_rate', 0.001), ('wd', 0.0001), ('beta1', 0.9),
```
Comment (anirudh2290) [inline example/sparse/factorization_machine/train.py]: Should we add these hyperparameters to user passed arguments with defaults ?

File: example/sparse/factorization_machine/data.py (hunk 2 above)
Comment (eric-haibin-lin) [reply example/sparse/factorization_machine/data.py]: I mentioned the size of the dataset in README. I can also add a warning here, too

File: example/sparse/factorization_machine/train.py (hunk 3 above)
Comment (eric-haibin-lin) [reply example/sparse/factorization_machine/train.py]: Yeah.. there're around 10 hyper-params..

**Comment by eric-haibin-lin** [comment]
+ @szha

File: example/sparse/factorization_machine/data.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    def __init__(self, real_iter):
+        super(DummyIter, self).__init__()
+        self.real_iter = real_iter
+        self.provide_data = real_iter.provide_data
+        self.provide_label = real_iter.provide_label
+        self.batch_size = real_iter.batch_size
+
+        for batch in real_iter:
+            self.the_batch = batch
+            break
```
Comment (szha) [inline example/sparse/factorization_machine/data.py]: Since real_iter is iterable, `self.the_batch = next(real_iter)` should do.

File: example/sparse/factorization_machine/data.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+
+def get_criteo_data(data_dir):
+    if not os.path.isdir(data_dir):
+        os.mkdir(data_dir)
+    try:
+        logging.info("Downloading dataset criteo to " + data_dir + " now ... ")
+        logging.info("This requires ~13GB disk space. Press control + C to stop downloading")
+        process = subprocess.Popen(['aws', 's3', 'cp', '--recursive', '--no-sign-request', \
+                                    's3://sparse-dataset/criteo', data_dir])
```
Comment (szha) [inline example/sparse/factorization_machine/data.py]: Is the bucket owned by us? Are we allowed to redistribute the data? The [user agreement from criteo labs](http://labs.criteo.com/2014/02/dataset/) say we can't distribute it.

File: example/sparse/factorization_machine/factorization_machine_model.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+    # interactions
+    w2 = mx.symbol.dot(x, v)
+    w2_squared = 0.5 * mx.symbol.square(data=w2)
+
+    # putting everything together
+    w_all = mx.symbol.Concat(w1, w2_squared, dim=1)
+    sum1 = mx.symbol.sum(data=w_all, axis=1, keepdims=True)
+    sum2 = 0.5 * mx.symbol.negative(bd_sum)
+    model = mx.sym.elemwise_add(sum1, sum2)
```
Comment (szha) [inline example/sparse/factorization_machine/factorization_machine_model.py]: Consider using fluent API to simplify code.

File: example/sparse/linear_classification/get_data.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import os, gzip
+import sys
+import mxnet as mx
+
+class DummyIter(mx.io.DataIter):
+    "A dummy iterator that always return the same batch, used for speed testing"
```
Comment (szha) [inline example/sparse/linear_classification/get_data.py]: Is there a way to avoid duplicate of this iterator? Would sym-link work?

**Comment by eric-haibin-lin** [comment]
Updated with fixes to address review comments. Please review again. @szha @anirudh2290

File: example/sparse/factorization_machine/train.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    args = parser.parse_args()
+    logging.info(args)
+    num_epoch = args.num_epoch
+    batch_size = args.batch_size
+    kvstore = args.kvstore
+    factor_size = args.factor_size
+    num_features = args.input_size
+    log_interval = args.log_interval
+    assert(args.data_train is not None)
+    assert(args.data_test is not None)
```
Comment (anirudh2290) [inline example/sparse/factorization_machine/train.py]: Can we add messages for the assert statements here ?

File: example/sparse/matrix_factorization/data.py [hunk 9]
```diff
@@ -18,38 +18,9 @@
...
-    if (not os.path.exists(data_name)):
-        print("Dataset " + data_name + " not present. Downloading now ...")
-        import urllib
-        zippath = os.path.join(data_dir, data_name + ".bz2")
-        urllib.urlretrieve(url + data_name + ".bz2", zippath)
-        os.system("bzip2 -d %r" % data_name + ".bz2")
-        print("Dataset " + data_name + " is now present.")
-    os.chdir("..")
+curr_path = os.path.dirname(os.path.abspath(os.path.expanduser(__file__)))
+sys.path.append(os.path.join(curr_path, "../../../tests/python/common"))
```
Comment (anirudh2290) [inline example/sparse/matrix_factorization/data.py]: Nitpick: This creates a dependency on the other test modules in repo. It would be nice if end users could just download the examples and run it on a mxnet installed system.

Comment (eric-haibin-lin) [reply example/sparse/matrix_factorization/data.py]: I don't know what's a good solution. Functions such as `GetMNIST_ubyte` is used in many examples, do we really want to have separate copies of this function in each example? But your concern is valid, I've seen users posting issues regarding the path problem. 
@szha

Comment (anirudh2290) [reply example/sparse/matrix_factorization/data.py]: If it is a function that is going to be used by many examples shouldn't it go with the python modules in mxnet python package. For eg. inside test_utils?

Comment (szha) [reply example/sparse/matrix_factorization/data.py]: Maybe move them to test_utils? @eric-haibin-lin

File: example/sparse/factorization_machine/train.py (hunk 8 above)
Comment (eric-haibin-lin) [reply example/sparse/factorization_machine/train.py]: sure

File: example/sparse/matrix_factorization/data.py (hunk 9 above)
Comment (eric-haibin-lin) [reply example/sparse/matrix_factorization/data.py]: sounds good. will do
